### DISCUSSION TITLE
revalidatePath not working with useSWR

### DISCUSSION THREAD

**Discussion by whoismaruf** [vercel/next.js#63865] [Help]
### Summary

I have a candidate table component that fetches candidates through the `useSWR` and nextjs's API route handler. Here is the code for fetching the data - 
```ts
# route.ts

import { NextRequest, NextResponse } from "next/server";
import fetcher from "@/utils/fetch-data";

export async function GET(
    req: NextRequest,
    { params }: { params: { id: number } }
) {
    const queryParams = req.nextUrl.searchParams;
    const campaign_id = params.id;
    const page = queryParams.get("page");
    const responses = await fetcher(`campaigns/${campaign_id}/candidate/?page=${page}`);
    return Response.json(responses);
}
```

And, it works fine.

Well, I have a button on my table that adds candidates via formAction. Here is the code for form action -

```ts
# action.ts

export async function createCandidateAction(prevState: any, formData: FormData ) {
    const session = await authSession();
    const campaign_id = parseInt(formData.get('campaign') as string);
    const response = await fetch(`${process.env.NEXTAUTH_BACKEND_URL}campaigns/${campaign_id}/candidate/`, {
        method: 'POST',
        body: formData,
        headers: {
            'Authorization': `Token ${session.key}`,
        }
    });
    if (response.status === 201) {
        const candidate = await response.json();
        revalidatePath(`/campaigns/${campaign_id}/candidates`);
        return candidate;
    } else {
        return {
                error: `🥹 Something went wrong! \n Error code: ${response.status}`
            };
    }
}
```
And, here is the issue. When I successfully add a candidate, it doesn't automatically revalidate the candidate list until I reload or change my focus and re-focus on the page. But before when I was using the regular `fetch` function, it worked flawlessly. To be more precise, the block of code where I checked `response.status===201` is running, as I have set the status 201 from the backend.

Additionally, here is the candidate table fetching code - 
```ts
# CampaignTable.tsx

export default function CandidatesTable({ id }: { id: number }) {
    const [page, setPage] = React.useState(1);
    const fetcher = (url: string) => axios.get(url).then(res => res.data)
    const { data, error, isLoading } = useSWR(`/api/campaigns/${id}/candidates?page=${page}`, fetcher)
    const pages = data?.total_pages
    const loadingState = (isLoading) ? "loading" : "idle";
    if (error) return <div className="text-center text-red-500">Failed to load data</div>
    return (
      # other code
   )
}
```

Thanks 🙌

**Comment by Amama-Fatima** [comment]
Hi. Have u found a solution?
I am encountering the exact same problem

**Reply by whoismaruf** [reply]
**Assalamualaykum, Fatima**. Yes, we have managed a solution, and what you need to do is, call the `mutate()` function inside the form action like this -

Say, you have a client component `<CandidateTable/>` which shows the table of candidates using `useSWR()`, modify your `useSWR()` like:

```ts
const { data, error, isLoading, mutate } = useSWR(`/api/candidates/`, fetcher, {
        revalidateOnMount: true,
    })
```

Now, on that same `<CandidateTable/>` component, on the top, say you have a modal component which creates new candidate, then pass the mutate as a prop to this component `<CandidateCreationFormModal mutate={mutate} />` and run the `mutate()` inside your form action which will re-validate the `data` you have on the `<CandidateTable/>` component : 

```ts
export default function CandidateCreationFormModal({
  mutate,
}: {
  mutate: Function;
}) {
   return(
     <form
          action={async (formData) => {
            const data = await createCandidateAction(null, formData);
            if (data?.error) {
              console.error(data.error);
            } else {
              mutate();
            }
          }}
        >
  )
}
```

I hope you can fix your issue soon inshallah 🎉
